File: 21_guessNumber/guess.java
package main;

import java.util.Random;
import java.util.Scanner;

public class Guess {
    // Guess Random Number Game
    // In Java

    // generated Number won't exceed this value
    static int max_num = 500;

    public static void main(String args[]) {
        Random ran = new Random();

        // Seeds the random generator
        ran.setSeed(System.currentTimeMillis());

        // Range, From & To, a Random Range Between 1, max_num
        int range_From = ran.nextInt(max_num);
        if (range_From == 0) range_From = 1;
        int range_To = ran.nextInt(max_num - range_From + 1) + range_From;

        // Random Number that the User have to guess :D
        int random_Num = ran.nextInt(range_To - range_From + 1) + range_From;

        // Number of allowed guesses` >= 3
        int number_Of_Guesses = 3 + (range_To - range_From) % 3;

        int user_attempts = 0; // Number of tries by User
        boolean user_won = false; // Is True if User enter random number
        String tip = ""; // Tip msg, changes according

        System.out.printf("Welcome to Guess the Random Number Game :D\n" +
            "Written in Java o_o\n The number is between {%d} ~ {%d}\n",
            range_From, range_To);

        // Used to read inputs from user
        Scanner reader = new Scanner(System.in);

        while (number_Of_Guesses > user_attempts) {
            String msg = "\r You have %d tries left,%s Enter a number: ";
            System.out.printf(msg, (number_Of_Guesses - user_attempts - 1),
            tip);

            int user_number = reader.nextInt();
            user_attempts += 1;
            if (user_number == random_Num) {
                user_won = true;
                break;
            } else if (user_number > random_Num)
                tip = "maybe you should try a smaller number :D";
            else if (user_number < random_Num)
                tip = "maybe you should try a bigger number :D";
        }

        if (user_won)
            System.out.printf("\nCongrats!!! YOU WON! with %d attempts\n",
            user_attempts);
        else
            System.out.printf("\n:( You Lost!, you literally had %d attempts," +
                " Number was %d \n", user_attempts, random_Num);
    }
}

File: 21_guessNumber/guess.go
package main

import (
	"bufio"
	"crypto/rand"
	"fmt"
	"math/big"
	"os"
	"strconv"
	"strings"
)

// Guess Random Number Game
// In Go

// generated Number won't exceed this value
const maxNum = 500

func main() {

	// Random Number Generator Function
	genRandom := func(from int64, to int64) int64 {
		num, err := rand.Int(rand.Reader, big.NewInt(to-from+1))
		if err != nil {
			fmt.Errorf("An Error Occured %v", err)
			os.Exit(1)
		}
		return num.Int64() + from
	}

	// Reader, read/get user input
	reader := bufio.NewReader(os.Stdin)

	// Range, From & To, a Random Range Between 1, max_num
	rangeFrom := genRandom(0, maxNum)
	rangeTo := genRandom(rangeFrom, maxNum)

	// Random Number that the User have to guess :D
	randomNum := genRandom(rangeFrom, rangeTo)

	// Number of allowed guesses` >= 3
	numberOfGuesses := 3 + (rangeTo-rangeFrom)%3

	var userAttempts int64 // Number of tries by User
	userWon := false       // Is True if User enter random number
	tip := ""              // Tip msg, changes according

	fmt.Printf("Welcome to Guess the Random Number Game :D\n"+
		"Written in Go o_o\n"+
		"The number is between %d ~ %d\n", rangeFrom, rangeTo)
	for numberOfGuesses > userAttempts {
		msg := "\r You have %d tries left,%v Enter a number: "
		fmt.Printf(msg, (numberOfGuesses - userAttempts - 1), tip)

		// User Input
		input, _ := reader.ReadString('\n')
		// Parse Input
		userNumber, _ := strconv.ParseInt(strings.TrimSpace(input), 0, 64)
		userAttempts++

		if userNumber == randomNum {
			userWon = true
			break
		} else if userNumber > randomNum {
			tip = "maybe you should try a smaller number :D"
		} else if userNumber < randomNum {
			tip = "maybe you should try a bigger number :D"
		}
	}

	if userWon {
		fmt.Printf("\nCongrats!!! YOU WON! with %d attempts\n", userAttempts)
	} else {
		fmt.Printf("\n:( You Lost!, you literally had %d attempts, Number was %d\n",
		userAttempts, randomNum)
	}
}
